File: DJIUXSDKWidgets/HorizontalVelocityWidget/DUXBetaHorizontalVelocityWidgetModel.m
//
//  DUXBetaHorizontalVelocityWidgetModel.m
//  DJIUXSDK
//
//  MIT License
//  
//  Copyright © 2018-2020 DJI
//  
//  Permission is hereby granted, free of charge, to any person obtaining a copy
//  of this software and associated documentation files (the "Software"), to deal
//  in the Software without restriction, including without limitation the rights
//  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
//  copies of the Software, and to permit persons to whom the Software is
//  furnished to do so, subject to the following conditions:

//  The above copyright notice and this permission notice shall be included in all
//  copies or substantial portions of the Software.

//  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
//  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
//  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
//  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
//  SOFTWARE.
//  

#import "DUXBetaHorizontalVelocityWidgetModel.h"
#import "DUXBetaBaseWidgetModel+Protected.h"
@import DJIUXSDKCore;

@interface DUXBetaHorizontalVelocityWidgetModel()

@property (nonatomic) DJISDKVector3D *velocityVector;
@property (nonatomic) NSMeasurement *horizontalVelocity;

@end

@implementation DUXBetaHorizontalVelocityWidgetModel

@synthesize horizontalVelocityUnits = _horizontalVelocityUnits;

- (instancetype)init {
    self = [super init];
    if (self) {
        self.velocityVector = nil;
        if (self.unitSystem == DUXBetaUnitSystemMetric) {
            self.horizontalVelocity = [[NSMeasurement alloc] initWithDoubleValue:0 unit: NSUnitLength.meters];
        } else {
            self.horizontalVelocity = [[NSMeasurement alloc] initWithDoubleValue:0 unit: NSUnitLength.feet];
        }
    }
    return self;
}

- (void)inSetup {
    BindSDKKey([DJIFlightControllerKey keyWithParam:DJIFlightControllerParamVelocity], velocityVector);
    
    BindRKVOModel(self, @selector(updateVelocityVector), velocityVector);
}

- (void)inCleanup {
    UnBindSDK;
    UnBindRKVOModel(self);
}

- (void)updateVelocityVector {
    if (self.velocityVector == nil) { return; }
    
    double velocityValue = sqrtf(self.velocityVector.x*self.velocityVector.x + self.velocityVector.y*self.velocityVector.y);
    NSMeasurement *horizontalVelocity = [[NSMeasurement alloc] initWithDoubleValue:velocityValue unit: NSUnitLength.meters];
    self.horizontalVelocity = [horizontalVelocity measurementByConvertingToUnit: self.horizontalVelocityUnits];
}

- (NSUnitLength *)horizontalVelocityUnits {
    if (_horizontalVelocityUnits) {
        return _horizontalVelocityUnits;
    } else if (self.unitSystem == DUXBetaUnitSystemMetric) {
        return NSUnitLength.meters;
    } else {
        return NSUnitLength.feet;
    }
}

- (void)setHorizontalVelocityUnits:(NSUnitLength *)horizontalVelocityUnits {
    _horizontalVelocityUnits = horizontalVelocityUnits;
    [self updateVelocityVector];
}

@end

File: UXSDKCore/UXSDKCore/Widget/Telemetry/HorizontalVelocity/DUXBetaHorizontalVelocityWidgetModel.swift
//
//  DUXBetaHorizontalVelocityWidgetModel.swift
//  UXSDKCore
//
//  MIT License
//  
//  Copyright © 2020 DJI
//  
//  Permission is hereby granted, free of charge, to any person obtaining a copy
//  of this software and associated documentation files (the "Software"), to deal
//  in the Software without restriction, including without limitation the rights
//  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
//  copies of the Software, and to permit persons to whom the Software is
//  furnished to do so, subject to the following conditions:

//  The above copyright notice and this permission notice shall be included in all
//  copies or substantial portions of the Software.

//  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
//  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
//  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
//  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
//  SOFTWARE.
//

import Foundation

/**
 * Widget Model for the DUXBetaHorizontalVelocityWidget used to define
 * the underlying logic and communication
 */
@objcMembers public class DUXBetaHorizontalVelocityWidgetModel: DUXBetaBaseWidgetModel {
    /// Value of the horizontal state of the aircraft
    dynamic public var horizontalVelocityState: HorizontalVelocityState = HorizontalVelocityState()
    /// The module handling the unit type related logic
    dynamic public var unitModule = DUXBetaUnitTypeModule()
    
    dynamic var velocityVector: DJISDKVector3D = DJISDKVector3D()
    
    public override init() {
        super.init()
        add(unitModule)
    }
    
    override public func inSetup() {
        if let key = DJIFlightControllerKey(param: DJIFlightControllerParamVelocity) {
            bindSDKKey(key, (\DUXBetaHorizontalVelocityWidgetModel.velocityVector).toString)
        }
        
        bindRKVOModel(self, #selector(updateIsConnected), (\DUXBetaHorizontalVelocityWidgetModel.isProductConnected).toString)
        bindRKVOModel(self, #selector(updateVelocity),
                      (\DUXBetaHorizontalVelocityWidgetModel.unitModule.unitType).toString,
                      (\DUXBetaHorizontalVelocityWidgetModel.velocityVector).toString,
                      (\DUXBetaHorizontalVelocityWidgetModel.isProductConnected).toString)
    }
    
    override public func inCleanup() {
        unbindSDK(self)
        unbindRKVOModel(self)
    }
    
    func updateVelocity() {
        //Compute system unit
        let unit = (unitModule.unitType == .Metric ? UnitLength.meters : UnitLength.feet)
        
        //Compute velocity value
        let velocity = sqrt(velocityVector.x * velocityVector.x + velocityVector.y * velocityVector.y)
        
        //Convert sdk value into corresponding measurement values based on system unit
        let horizontalVelocity = NSMeasurement(doubleValue: velocity, unit: UnitLength.meters).converting(to: unit)
        
        //Update altitude model state
        horizontalVelocityState.currentVelocity = horizontalVelocity
    }
    
    func updateIsConnected() {
        horizontalVelocityState.isProductDisconnected = !isProductConnected
    }
}

/**
 * Class to represent states of the horizontal velocity
 */
@objcMembers public class HorizontalVelocityState: NSObject {
    /// Product is disconnected.
    public dynamic var isProductDisconnected: Bool = false
    /// The current velocity when aircraft is moving horizontally
    public dynamic var currentVelocity: Measurement<Unit> = .init(value: 0.0, unit: UnitLength.meters)
}
